server/state/in_memory_state.go

```diff
@@ -88,47 +124,76 @@ func (s *InMemoryState) HasSandbox(id string) bool {
 }
 
 // DeleteSandbox removes a sandbox from the state
-func (s *InMemoryState) DeleteSandbox(id string) error {
+func (s *InMemoryState) DeleteSandbox(id string) (err error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
 	if _, exist := s.sandboxes[id]; !exist {
-		return fmt.Errorf("no sandbox with ID %v exists, cannot delete", id)
+		return errNoSuchSandbox(id)
 	}
 
-	name := s.sandboxes[id].Name()
-	containers := s.sandboxes[id].Containers()
-	infraContainer := s.sandboxes[id].InfraContainer()
+	sb := s.sandboxes[id]
 
-	delete(s.sandboxes, id)
-	s.podNameIndex.Release(name)
-	if err := s.podIDIndex.Delete(id); err != nil {
-		return fmt.Errorf("error unregistering sandbox ID: %v", err)
+	if err := s.deleteSandboxMappings(sb); err != nil {
+		return err
 	}
+	defer func() {
+		if err != nil {
+			if err2 := s.addSandboxMappings(sb); err2 != nil {
+				logrus.Errorf("Error re-adding sandbox mappings: %v", err2)
+			}
+		}
+	}()
 
+	removedCtrs := make([]*oci.Container, 0, len(sb.Containers()))
 	// If there are containers left in the sandbox delete them from the mappings
-	for _, ctr := range containers {
+	for _, ctr := range sb.Containers() {
 		if err := s.deleteContainerMappings(ctr, true); err != nil {
 			return fmt.Errorf("error removing container %v mappings: %v", ctr.ID(), err)
 		}
-	}
+		removedCtrs = append(removedCtrs, ctr)
```

Comment:
Same here, you return from error before defer:in to re-add the stuff in removedCtrs. Just drop the array and do a defer per container.